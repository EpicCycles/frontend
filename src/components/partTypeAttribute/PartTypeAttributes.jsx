import React from "react";
import PartTypeAttributeEdit from "./PartTypeAttributeEdit";
import {
    moveObjectDownOnePlace,
    moveObjectToBottom,
    moveObjectToTop,
    moveObjectUpOnePlace} from "../framework/helpers/framework";
import FrameworkMoves from "../framework/FrameworkMoves";
import {findIndexOfObjectWithKey} from "../../helpers/utils";
import {NEW_ELEMENT_ID} from "../../helpers/constants";
import {getModelKey} from "../app/model/helpers/model";
import {NEW_ATTRIBUTE} from "../partType/helpers/partType";

class PartTypeAttributes extends React.Component {
    handleAttributeChange = (attributeKey, updatedAttribute) => {
        const attributesWithUpdates = this.props.attributes.slice();
        const attributeToUpdateIndex = findIndexOfObjectWithKey(attributesWithUpdates, attributeKey);
        if (attributeToUpdateIndex > -1) {
            attributesWithUpdates[attributeToUpdateIndex] = updatedAttribute;
        } else {
            attributesWithUpdates.push(updatedAttribute);
        }
        this.props.handlePartTypeChange(`attributes_${this.props.partTypeKey}`, attributesWithUpdates);
    };
    moveUp = (fieldName) => {
        const fields = fieldName.split('_');
        const attributeKey = fields[1];
        this.props.handlePartTypeChange(`attributes_${this.props.partTypeKey}`, moveObjectUpOnePlace(this.props.attributes, attributeKey));
    };
    moveDown = (fieldName) => {
        const fields = fieldName.split('_');
        const attributeKey = fields[1];
        this.props.handlePartTypeChange(`attributes_${this.props.partTypeKey}`, moveObjectDownOnePlace(this.props.attributes, attributeKey));
    };
    moveToTop = (fieldName) => {
        const fields = fieldName.split('_');
        const attributeKey = fields[1];
        this.props.handlePartTypeChange(`attributes_${this.props.partTypeKey}`, moveObjectToTop(this.props.attributes, attributeKey));
    };
    moveToBottom = (fieldName) => {
        const fields = fieldName.split('_');
        const attributeKey = fields[1];
        this.props.handlePartTypeChange(`attributes_${this.props.partTypeKey}`, moveObjectToBottom(this.props.attributes, attributeKey));
    };

    render() {
        const { partTypeKey, attributes } = this.props;
        const attributesToUse = attributes ? attributes.filter(attribute => !(attribute.delete || (attribute.dummyKey === NEW_ELEMENT_ID))) : [];
        const newAttributes = attributes ? attributes.filter(attribute => (attribute.dummyKey === NEW_ELEMENT_ID)) : [];
        let newAttributeDisplay = (newAttributes.length > 0) ? newAttributes[0] : NEW_ATTRIBUTE;
        return <table key={`attributes_${partTypeKey}`} className="full">
            <tbody>
            <tr>
                <th>Attribute Details</th>
                <th>Options</th>
                <th>Position</th>
            </tr>
            {attributesToUse.map((attribute) => {
                const componentKey = getModelKey(attribute);
                const className = attribute.error ? "error" : "";
                const rowTitle = attribute.error ? attribute.error_detail : "";
                return (
                    <tr
                        key={`attributeRow${componentKey}`}
                        className={className}
                        title={rowTitle}
                    >
                        <PartTypeAttributeEdit
                            key={`attributeEdit${componentKey}`}
                            attribute={attribute}
                            componentKey={componentKey}
                            partType={partTypeKey}
                            handleAttributeChange={this.handleAttributeChange}
                        />
                        <td>
                            {attributesToUse.length > 1 &&
                            <FrameworkMoves
                                componentKey={componentKey}
                                moveToTop={this.moveToTop}
                                moveUp={this.moveUp}
                                moveDown={this.moveDown}
                                moveToBottom={this.moveToBottom}
                            />
                            }
                        </td>
                    </tr>
                );
            })}
            <tr key={`newattributeRow${partTypeKey}`}>
                <PartTypeAttributeEdit
                    key="attributeEditNew"
                    attribute={newAttributeDisplay}
                    componentKey={NEW_ELEMENT_ID}
                    handleAttributeChange={this.handleAttributeChange}
                />
                <td/>
            </tr>
            </tbody>
        </table>;
    }
}

export default PartTypeAttributes;