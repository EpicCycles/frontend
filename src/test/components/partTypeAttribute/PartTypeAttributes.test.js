import React from 'react';
import toJson from 'enzyme-to-json';
import PartTypeAttributes from '../../../components/partTypeAttribute/PartTypeAttributes';

describe('PartTypeAttributes tests', () => {
  const attributes = [
    {
      id: 12,
      options: [],
      attribute_name: 'Colour',
      in_use: true,
      mandatory: true,
      placing: 1,
      attribute_type: '1',
      partType: 18,
    },
    {
      id: 17,
      options: [],
      attribute_name: 'Size',
      in_use: true,
      mandatory: true,
      placing: 2,
      attribute_type: '1',
      partType: 18,
    },
    {
      id: 11,
      options: [
        {
          id: 5,
          attribute_option: 'Long',
          part_type_attribute: 11,
        },
        {
          id: 4,
          attribute_option: 'Medium',
          part_type_attribute: 11,
        },
        {
          id: 3,
          attribute_option: 'Short',
          part_type_attribute: 11,
        },
      ],
      attribute_name: 'Valve Length',
      in_use: true,
      mandatory: false,
      placing: 1,
      attribute_type: '4',
      partType: 19,
    },
  ];

  const partTypeId = 234;

  it('renders the form correctly with no attributes', () => {
    const input = shallow(<PartTypeAttributes attributes={[]} partTypeId={partTypeId} />);
    expect(toJson(input)).toMatchSnapshot();
  });
  it('renders the form correctly with attributes', () => {
    const input = shallow(
      <PartTypeAttributes attributes={attributes} partTypeId={attributes[0].partType} />,
    );
    expect(toJson(input)).toMatchSnapshot();
  });
  it('renders the form correctly when all attributes deleted', () => {
    const attributesDeleted = [
      { id: 123, partType: 101, delete: true },
      { id: 323, partType: 101, delete: true },
      { partType: 1, dummyKey: 'hjgfkuyg34', delete: true },
    ];
    const input = shallow(
      <PartTypeAttributes
        attributes={attributesDeleted}
        partTypeId={attributesDeleted[0].partType}
      />,
    );
    expect(toJson(input)).toMatchSnapshot();
  });
});
