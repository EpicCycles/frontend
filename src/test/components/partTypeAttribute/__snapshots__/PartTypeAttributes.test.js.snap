// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartTypeAttributes tests renders the form correctly when all attributes deleted 1`] = `
<table
  className="full"
  key="attributes_undefined"
>
  <tbody>
    <tr>
      <th>
        Attribute Details
      </th>
      <th>
        Options
      </th>
      <th>
        Position
      </th>
    </tr>
    <tr
      key="newattributeRowundefined"
    >
      <PartTypeAttributeEdit
        attribute={
          Object {
            "attribute_name": "",
            "attribute_type": 1,
            "in_use": true,
            "mandatory": false,
          }
        }
        componentKey="new"
        handleAttributeChange={[Function]}
        key="attributeEditNew"
      />
      <td />
    </tr>
  </tbody>
</table>
`;

exports[`PartTypeAttributes tests renders the form correctly with attributes 1`] = `
<table
  className="full"
  key="attributes_undefined"
>
  <tbody>
    <tr>
      <th>
        Attribute Details
      </th>
      <th>
        Options
      </th>
      <th>
        Position
      </th>
    </tr>
    <tr
      className=""
      key="attributeRow12"
      title=""
    >
      <PartTypeAttributeEdit
        attribute={
          Object {
            "attribute_name": "Colour",
            "attribute_type": "1",
            "id": 12,
            "in_use": true,
            "mandatory": true,
            "options": Array [],
            "partType": 18,
            "placing": 1,
          }
        }
        componentKey={12}
        handleAttributeChange={[Function]}
        key="attributeEdit12"
      />
      <td>
        <FrameworkMoves
          componentKey={12}
          moveDown={[Function]}
          moveToBottom={[Function]}
          moveToTop={[Function]}
          moveUp={[Function]}
        />
      </td>
    </tr>
    <tr
      className=""
      key="attributeRow17"
      title=""
    >
      <PartTypeAttributeEdit
        attribute={
          Object {
            "attribute_name": "Size",
            "attribute_type": "1",
            "id": 17,
            "in_use": true,
            "mandatory": true,
            "options": Array [],
            "partType": 18,
            "placing": 2,
          }
        }
        componentKey={17}
        handleAttributeChange={[Function]}
        key="attributeEdit17"
      />
      <td>
        <FrameworkMoves
          componentKey={17}
          moveDown={[Function]}
          moveToBottom={[Function]}
          moveToTop={[Function]}
          moveUp={[Function]}
        />
      </td>
    </tr>
    <tr
      className=""
      key="attributeRow11"
      title=""
    >
      <PartTypeAttributeEdit
        attribute={
          Object {
            "attribute_name": "Valve Length",
            "attribute_type": "4",
            "id": 11,
            "in_use": true,
            "mandatory": false,
            "options": Array [
              Object {
                "attribute_option": "Long",
                "id": 5,
                "part_type_attribute": 11,
              },
              Object {
                "attribute_option": "Medium",
                "id": 4,
                "part_type_attribute": 11,
              },
              Object {
                "attribute_option": "Short",
                "id": 3,
                "part_type_attribute": 11,
              },
            ],
            "partType": 19,
            "placing": 1,
          }
        }
        componentKey={11}
        handleAttributeChange={[Function]}
        key="attributeEdit11"
      />
      <td>
        <FrameworkMoves
          componentKey={11}
          moveDown={[Function]}
          moveToBottom={[Function]}
          moveToTop={[Function]}
          moveUp={[Function]}
        />
      </td>
    </tr>
    <tr
      key="newattributeRowundefined"
    >
      <PartTypeAttributeEdit
        attribute={
          Object {
            "attribute_name": "",
            "attribute_type": 1,
            "in_use": true,
            "mandatory": false,
          }
        }
        componentKey="new"
        handleAttributeChange={[Function]}
        key="attributeEditNew"
      />
      <td />
    </tr>
  </tbody>
</table>
`;

exports[`PartTypeAttributes tests renders the form correctly with no attributes 1`] = `
<table
  className="full"
  key="attributes_undefined"
>
  <tbody>
    <tr>
      <th>
        Attribute Details
      </th>
      <th>
        Options
      </th>
      <th>
        Position
      </th>
    </tr>
    <tr
      key="newattributeRowundefined"
    >
      <PartTypeAttributeEdit
        attribute={
          Object {
            "attribute_name": "",
            "attribute_type": 1,
            "in_use": true,
            "mandatory": false,
          }
        }
        componentKey="new"
        handleAttributeChange={[Function]}
        key="attributeEditNew"
      />
      <td />
    </tr>
  </tbody>
</table>
`;
